### PR TITLE
23 implement induction machine model

### LINKED ISSUE #23: Implement induction machine model

### PR DESCRIPTION
**Pull request by artturuusila**
This PR adds a model for the induction machine and a control system for the stator current.

closes #23

### CODE REVIEW

File: soft4pes/control/mpc/controllers/im_mpc_curr_ctr.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+        # Predict the reference by rotating the current reference
+        y_ref = np.zeros((self.Np, 2))
+        i_ref_temp = i_ref
+        for k in range(self.Np):
+            y_ref[k, :] = np.dot(R_ref, i_ref_temp)
+            i_ref_temp = y_ref[k, :]
+
+        # Solve the control problem
+        uk = self.solver(sys, conv, self, y_ref)
```
Comment (PetPKar) [inline soft4pes/control/mpc/controllers/im_mpc_curr_ctr.py:117]: Given the above comments, I am thinking of structure like this:

* One file for the outer loop: This file generates the current reference (for machine or grid)
* One file with the FCS-MPC current control problem: 
-->We feed into this file the current reference
--> We generate the reference trajectory in the horizon (as you do here)
--> We call the possible way to solve the problem
* One file for the solver

File: soft4pes/control/mpc/controllers/im_mpc_curr_ctr.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+        Returns
+        -------
+        1 x 2 ndarray of floats
+            The next state of the system [p.u.] (step k+1).
+        """
+
+        return np.dot(self.state_space.A, xk) + np.dot(self.state_space.B, uk)
+
+    def save_data(self, iS_ref, u_k, T_ref, t):
```
Comment (PetPKar) [inline soft4pes/control/mpc/controllers/im_mpc_curr_ctr.py:147]: I assume there will be more quantities that will be stored for plotting (actual currents, the applied switch positions, flux)

**Review by PetPKar — COMMENTED** [review]
some comments on structure and small parts of the code

File: soft4pes/control/mpc/controllers/im_mpc_curr_ctr.py [hunk 3]
```diff
@@ -130,16 +130,16 @@ def get_next_state(self, sys, xk, uk, k):
         sys : system object
             The system object, not used in this method.
         xk : 1 x 2 ndarray of floats
-            The current state of the system [p.u.].
+            The current state of the system [p.u.] (step k).
         uk : 1 x 3 ndarray of ints
-            Converter 3-phase switch position.
+            Converter three-phase switch position.
```
Comment (artturuusila) [inline soft4pes/control/mpc/controllers/im_mpc_curr_ctr.py:135]: In other files, I have used 3-phase (not three-phase). Should this be changed?

File: soft4pes/model/machine/induction_machine.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+        F = np.array([[-1 / tauS, 0, Xm / (tauR * D), wr * Xm / D],
+                      [0, -1 / tauS, -wr * Xm / D, Xm / (tauR * D)],
+                      [Xm / tauR, 0, -1 / tauR, -wr],
+                      [0, Xm / tauR, wr, -1 / tauR]])
+
+        G = Xr / D * np.dot(np.array([[1, 0], [0, 1], [0, 0], [0, 0]]),
+                            K) * v_dc / 2
+
+        A = np.eye(4) + F * Ts_pu
```
[the remark this answers is not in the document]
Comment (artturuusila) [reply soft4pes/model/machine/induction_machine.py:237]: Will be added in the future

File: soft4pes/model/machine/base_machine.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    npp: int
+
+    def __post_init__(self):
+        self.V = np.sqrt(2 / 3) * self.Vr
+        self.I = np.sqrt(2) * self.Ir
+        self.w = 2 * np.pi * self.fr
+        self.S = 3 / 2 * self.V * self.I
+        self.Z = self.V / self.I
+        self.L = self.Z / self.w
+        self.T = self.npp * self.S / self.w
```
[the remark this answers is not in the document]
Comment (artturuusila) [reply soft4pes/model/machine/base_machine.py]: Power factor added to the equation

**Comment by artturuusila** [comment]
@PetPKar PR comments addressed, please review/accept

**Comment by PetPKar** [comment]
Things look fine after the implemented changed

**Review by PetPKar — APPROVED** [review]

**Pull request merged**